### ISSUE TITLE
Isort violates line_length with include_trailing_commas setting in special cases

### ISSUE THREAD

**Issue by micahjsmith** [PyCQA/isort#1640]
It seems possible to get isort to violate `line_length` when using `include_trailing_commas`.

Consider the following example:

test.py:
```
from mylib import aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa, bbbbbbbbb
```

.isort.cfg:
```
[settings]
include_trailing_comma = True
line_length = 79
multi_line_output = 4
```

Then running isort produces

```
from mylib import (
    aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa, bbbbbbbbb,)
```

You will notice if you inspect this in your editor that the second line of the fixed file is 80 characters, violating line_length.

I would instead expect behavior where `bbbbbbbbb` is placed on line 3, such that no line is over the max length:
```
from mylib import (
    aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa,
    bbbbbbbbb,)
```

- isort version: 5.7.0
- python version: 3.7.7

Similar to #1634

**Comment by gofr** [comment]
Yeah, I mentioned this case in https://github.com/PyCQA/isort/issues/1634#issuecomment-757344253. My pull request there would solve both these issues (but still needs a bit of work like tests and docs).

As I understand it this behavior is intentional, but I don't really understand why it would be. :grimacing: Well, backwards-compatibility may be an issue.

The behavior was "introduced" in 92f53fc47 (to fix some other edge cases). It did not take into account the use of `include_trailing_comma`. The code was last changed for issue #654 which was I think really a partial fix for the `include_trailing_comma` bug.

**Comment by micahjsmith** [comment]
I looked over the other thread first, and yeah the example there is very similar but not the same. In this case, it actually does violate line_length, no? rather than just producing surprising behavior that may be acceptable.

Obviously related implementation issues and I'm sure that your PR there would resolve this issue also. Thanks for your work there.

**Comment by gofr** [comment]
Yes, you're right. It is a different issue than the one originally reported there. The issue there could be fixed by switching from mode 5 to mode 6. _This_ issue with mode 4 does violate `line_length` and can't be fixed by switching to another mode. It can only be worked around by not using `include_trailing_comma`.

In the single comment I linked to I did _mean_ the bug you reported, but I didn't include an example of it. :slightly_smiling_face:

I can re-work my fix from there if needed to only address _this_ part...

It may also be _intended_ that `include_trailing_comma` doesn't "work" with mode 4 like you're doing. If that's the case I'd like to know why. :stuck_out_tongue:

**Comment by timothycrosley** [comment]
> It may also be intended that include_trailing_comma doesn't "work" with mode 4 like you're doing. If that's the case I'd like to know why.

I don't think it's intended that it doesn't work - it's just a side effect that the vast majority of people don't like the asthetic of that particular combination and so the bug hasn't been caught yet. For the other issue, I think using the alternate output mode is the solution, for this mode I think it makes sense to fix it since I don't think anyone is intentionally or would be relying on going beynod the line-length so it won't cause formatting churn. If it does hopefully the integration tests in the project would catch it.

**Comment by gofr** [comment]
I'm going to have to continue to be annoying. :stuck_out_tongue:

The only way to fix this bug in mode 4 without also affecting modes 5/6 would be to add an additional parameter to the function that they share: https://github.com/PyCQA/isort/blob/20e0c6fe5ba39071f9cb04ef2a10f02309fa2157/isort/wrap_modes.py#L185
I think it'll make the code more confusing.

I'll see if I can create a few separate pull requests: tests for all these edge cases, a complete fix for just this #1640, and a fix for both this and #1634 (more complete than my earlier pull request). Maybe you misunderstand what I have in mind. I should have some time later this week. Then you can decide which one you like best.

**Comment by gofr** [comment]
... And now I also understand why you're reluctant to fix #1634. Deprecating a multi-line mode is pretty complicated. #1653 attempts to do it (fixing both #1634 and #1640). All tests pass, but I'm not super confident I didn't miss something. #1652 fixes just #1640.

**Comment by timothycrosley** [comment]
Closing as this has been fixed in develop thanks to the great work by @gofr :tada:
